yt/frontends/ramses/particle_handlers.py

```diff
@@ -0,0 +1,329 @@
+    """
+    This is a base class that on instantiation registers the file
+    handler into the list. Used as a metaclass.
+    """
+    def __new__(meta, name, bases, class_dict):
+        cls = type.__new__(meta, name, bases, class_dict)
+        if cls.ptype is not None:
+            register_particle_handler(cls)
+        return cls
+
+
+@add_metaclass(RAMSESParticleFileHandlerRegister)
+class ParticleFileHandler(object):
+    '''
+    Abstract class to handle particles in RAMSES. Each instance
+    represents a single file (one domain).
+
+    To add support to a new particle file, inherit from this class and
+    implement all functions containing a `NotImplementedError`.
+
+    See `SinkParticleFileHandler` for an example implementation.'''
+
+    # These properties are static properties
+    ptype = None  # The name to give to the particle type
+    fname = None  # The name of the file(s).
+    file_descriptor = None # The name of the file descriptor (if any)
+
+    attrs = None  # The attributes of the header
+    known_fields = None  # A list of tuple containing the field name and its type
+
+    # These properties are computed dynamically
+    field_offsets = None     # Mapping from field to offset in file
+    field_types = None       # Mapping from field to the type of the data (float, integer, ...)
+    local_particle_count = None  # The number of particle in the domain
+
+    def __init__(self, ds, domain_id):
+        '''
+        Initalize an instance of the class. This automatically sets
+        the full path to the file. This is not intended to be
+        overriden in most cases.
+
+        If you need more flexibility, rewrite this function to your
+        need in the inherited class.
+        '''
+        self.ds = ds
+        self.domain_id = domain_id
+        basename = os.path.abspath(
+              os.path.dirname(ds.parameter_filename))
+        iout = int(
+            os.path.basename(ds.parameter_filename)
+            .split(".")[0].
+            split("_")[1])
+        icpu = domain_id
+
+        self.fname = os.path.join(
+            basename,
+            self.fname.format(iout=iout, icpu=icpu))
+
+        if self.file_descriptor is not None:
+            self.file_descriptor = os.path.join(
+                basename,
+                self.file_descriptor)
+
+    @property
+    def exists(self):
+        '''
+        This function should return True if the *file* the instance
+        exists. It is called for each file of the type found on the
+        disk.
+
+        By default, it just returns whether the file exists. Override
+        it for more complex cases.
+        '''
+        return os.path.exists(self.fname)
+
+    @property
+    def has_part_descriptor(self):
+        '''
+        This function should return True if a *file descriptor*
+        exists.
+
+        By default, it just returns whether the file exists. Override
+        it for more complex cases.
+        '''
```

Comment:
In Python2 these docstrings won't get inherited by subclasses but it looks like that's been fixed in Python3. tbh it's probably not worth worrying about making sure the docstrings are correct on the subclasses in Python2 because it needs to happen manually every time you define a function that the superclass defines.

Reply:
That's annoying. I'm using Python3 so I never noticed it.
In any case, the docstring is more for devs than for users, so it's probably enough to have them in the code, even though it's not inherited correctly.